### PR TITLE
[docs] padding-free training

### PR DESCRIPTION
**Pull request by stevhliu**
adds a padding-free doc to discuss usage and caveats (see #46093 for more context)

### CODE REVIEW

File: docs/source/en/padding_free.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+specific language governing permissions and limitations under the License.
+
+⚠️ Note that this file is in Markdown but contain specific syntax for our doc-builder (similar to MDX) that may not be
+rendered properly in your Markdown viewer.
+
+-->
+
+# Padding-free training
+
+Padding-free training (also called packing) concatenates several samples into a single sequence instead of padding each one to a fixed length. The model needs to know where each sample ends so attention doesn't mix tokens across samples.
```
Comment (vasqu) [inline docs/source/en/padding_free.md]:
```suggestion
Padding-free training (also called packing) concatenates several samples into a single sequence instead of padding each one to a fixed length. The model needs to know where each sample ends so (linear) attention doesn't mix tokens across samples.
```

File: docs/source/en/padding_free.md [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+There are two ways to provide those boundaries.
+
+- Infer them from `position_ids` at runtime.
```
Comment (vasqu) [inline docs/source/en/padding_free.md]: Let's add a warning before that this is not ideal and not the preferred way - and only limited to pure attention models, not models like qwen3 next or qwen3_5 that use a flavor of linear attention (gated delta net (GFN))

File: docs/source/en/padding_free.md [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+- Prepare them ahead of time with a data collator.
+
+The recommended approach is the data collator. This guide explains why and covers the caveats of the `position_ids` path.
+
+> [!TIP]
+> Padding-free relies on a FlashAttention implementation. Install [flash-attn](https://github.com/Dao-AILab/flash-attention) and load the model with `attn_implementation="flash_attention_2"` (or `"flash_attention_3"`), since only the FlashAttention kernels expose the variable-length path that a flattened batch needs.
```
Comment (vasqu) [inline docs/source/en/padding_free.md]: Depending on whether it's linear attention, there are oftentimes other packages needed like mamba-ssm (with the causal-conv1d one) and/or flash-linear-attention. It depends on the specific original implementation which may used another kernel or another.

I would note that a lot of these packages have a fallback via `kernels` so in most instances kernels alone can work a lot of times already (minus flash-linear-attention I think)

File: docs/source/en/padding_free.md [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+import torch
+from datasets import load_dataset
+from transformers import AutoModelForCausalLM, AutoTokenizer, DataCollatorWithFlattening, Trainer, TrainingArguments
+
+model_id = "meta-llama/Llama-3.2-1B"
+tokenizer = AutoTokenizer.from_pretrained(model_id)
+model = AutoModelForCausalLM.from_pretrained(
+    model_id,
+    attn_implementation="flash_attention_2",
+    device_map="cuda",
```
Comment (vasqu) [inline docs/source/en/padding_free.md]:
```suggestion
    device_map="auto",
```
there exists a xpu and npu branch so I think we can keep it general

File: docs/source/en/padding_free.md [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    args=TrainingArguments(output_dir="padding-free-llama"),
+    train_dataset=dataset,
+    data_collator=data_collator,
+)
+trainer.train()
+```
+
+## Linear attention and convolution models
+
+Gated DeltaNet (GDN), other linear-attention layers, and causal convolutions have no `position_ids`-only path, by design. Preparing the data with the collator is the only supported option for these models.
```
Comment (vasqu) [inline docs/source/en/padding_free.md:84]: Ah ok here it is, I think we should mention this earlier because some may not read the full document see the first example and wonder why things dont work as expected

File: docs/source/en/padding_free.md [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+```python
+from transformers import DataCollatorWithFlattening
+
+data_collator = DataCollatorWithFlattening(
+    return_flash_attn_kwargs=True,
+    return_seq_idx=True,
+)
+```
+
+These models also need their kernels installed, [flash-linear-attention](https://github.com/fla-org/flash-linear-attention) for the linear-attention recurrence and [causal-conv1d](https://github.com/Dao-AILab/causal-conv1d) for the convolution. Without them, the model falls back to reference implementations that ignore the boundary kwargs and mix tokens across samples.
```
Comment (vasqu) [inline docs/source/en/padding_free.md]: I think bamba also has support which relies on mamba-ssm :D its messy sorry

**Review by vasqu — COMMENTED** [review]
Some initial comments, I think we should mention a few things earlier and warn directly there but other than that super clean

File: docs/source/en/padding_free.md (hunk 5 above)
Comment (stevhliu) [reply docs/source/en/padding_free.md:84]: 👍 i added a warning earlier on with a link to this section so it'll stand out and flipped the `position_ids` and data collator sections so the recommended path comes first

File: docs/source/en/padding_free.md [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+## Prepare boundaries with a data collator
+
+Preparing the boundary kwargs up front removes the problems above and behaves identically whether or not you compile.
+
+Use [`DataCollatorWithFlattening`] to flatten each batch and return the boundary information. Set `return_flash_attn_kwargs=True` so the collator precomputes the boundaries instead of leaving them to be inferred from `position_ids` at runtime. Pass it to [`Trainer`] and don't add an `attention_mask`, since the flattened batch already encodes the boundaries and a mask conflicts with the packed layout.
+
+> [!TIP]
+> Padding-free relies on a FlashAttention implementation for standard attention models. Install [flash-attn](https://github.com/Dao-AILab/flash-attention) and load the model with `attn_implementation="flash_attention_2"` (or `"flash_attention_3"`), since only the FlashAttention kernels expose the variable-length path that a flattened batch needs.
```
Comment (vasqu) [inline docs/source/en/padding_free.md]: I think we can also advise to use hf kernels which can also work as fallback then (feel like a lot are still not aware and its a good opportunity to promote)

**Review by vasqu — APPROVED** [review]
Just one small suggestion to add kernels in the FA path as well but otherwise LGTM 🫡
